File: forge/features/platform/templates/webhooks/python/files/src/app/services/webhook_service.py
"""Outbound webhook delivery with HMAC-SHA256 signing.

Each delivery carries three headers the receiver MUST verify:

  * ``X-Webhook-Timestamp``  — RFC 3339 / Unix seconds; reject if older
    than ~5 minutes to limit the replay window.
  * ``X-Webhook-Nonce``      — 128-bit UUID; reject if seen before in the
    freshness window (maintain a short-TTL cache keyed by nonce).
  * ``X-Webhook-Signature``  — ``HMAC-SHA256(secret, timestamp "." nonce
    "." body)``, hex digest.

The nonce closes the within-same-second replay window that a
timestamp-only HMAC leaves open. Delivery is in-process, best-effort —
pair with ``background_tasks`` for retry semantics.
"""

from __future__ import annotations

import fnmatch
import hashlib
import hmac
import ipaddress
import json
import logging
import os
import socket
import time
import uuid
from typing import Any
from urllib.parse import urlparse

from app.data.models.webhook import Webhook
from app.domain.webhook import WebhookDeliveryResult

logger = logging.getLogger(__name__)


class WebhookUrlError(ValueError):
    """Raised when a webhook target URL is rejected by the SSRF guard."""


def _is_dev() -> bool:
    env = os.environ.get("ENV", os.environ.get("ENVIRONMENT", "production")).strip().lower()
    return env in {"development", "dev", "test", "testing", "local", "ci"}


def _ip_is_blocked(ip: ipaddress.IPv4Address | ipaddress.IPv6Address) -> bool:
    """Block any address that could reach internal infrastructure: loopback,
    RFC1918 / unique-local private ranges, link-local (incl. the 169.254.169.254
    cloud-metadata endpoint), multicast, reserved, and the unspecified
    address. IPv4-mapped IPv6 is unwrapped first so ``::ffff:127.0.0.1`` can't
    sneak past."""
    if isinstance(ip, ipaddress.IPv6Address) and ip.ipv4_mapped is not None:
        ip = ip.ipv4_mapped
    return (
        ip.is_private
        or ip.is_loopback
        or ip.is_link_local
        or ip.is_multicast
        or ip.is_reserved
        or ip.is_unspecified
    )


def validate_outbound_url(url: str) -> None:
    """Reject webhook targets that point at internal/non-public hosts (SSRF).

    Enforced at BOTH create time (fast feedback) and fire time (the
    authoritative check, since DNS can change between the two). Requires an
    ``https`` scheme in a production posture; ``http`` is tolerated only in a
    dev/test environment. Resolves the host and blocks if ANY resolved
    address is internal. Raises :class:`WebhookUrlError` on rejection.
    """
    parsed = urlparse(url)
    scheme = (parsed.scheme or "").lower()
    if scheme not in {"http", "https"}:
        raise WebhookUrlError(f"unsupported URL scheme {scheme!r}; use https")
    if scheme == "http" and not _is_dev():
        raise WebhookUrlError("http webhook targets are not allowed in production; use https")
    host = parsed.hostname
    if not host:
        raise WebhookUrlError("webhook URL has no host")

    # A literal IP bypasses DNS — check it directly. Otherwise resolve every
    # address the host maps to and block if any is internal.
    candidates: list[ipaddress.IPv4Address | ipaddress.IPv6Address] = []
    try:
        candidates.append(ipaddress.ip_address(host))
    except ValueError:
        try:
            infos = socket.getaddrinfo(host, parsed.port or (443 if scheme == "https" else 80))
        except socket.gaierror as e:
            raise WebhookUrlError(f"could not resolve webhook host {host!r}") from e
        for info in infos:
            addr = info[4][0]
            try:
                candidates.append(ipaddress.ip_address(addr.split("%", 1)[0]))
            except ValueError:
                continue
    if not candidates:
        raise WebhookUrlError(f"webhook host {host!r} resolved to no usable address")
    for ip in candidates:
        if _ip_is_blocked(ip):
            raise WebhookUrlError(
                f"webhook host {host!r} resolves to a non-public address ({ip}); refused"
            )


def _sign(secret: str, body: bytes, timestamp: str, nonce: str) -> str:
    """Return the hex digest of ``HMAC-SHA256(secret, timestamp.nonce.body)``.

    Nonce is included in the signed message so receivers cannot accept a
    replayed (timestamp, body) pair without also matching the nonce — and
    the nonce is expected to be unique per delivery.
    """
    message = timestamp.encode() + b"." + nonce.encode() + b"." + body
    digest = hmac.new(secret.encode(), message, hashlib.sha256).hexdigest()
    return digest


def matches_event(webhook: Webhook, event: str) -> bool:
    """An empty ``events`` list means subscribe-all; otherwise fnmatch."""
    if not webhook.events:
        return True
    return any(fnmatch.fnmatch(event, pattern) for pattern in webhook.events)


async def deliver(webhook: Webhook, event: str, payload: dict[str, Any]) -> WebhookDeliveryResult:
    """POST ``payload`` to ``webhook.url`` with an HMAC signature header.

    Returns a ``WebhookDeliveryResult`` regardless of outcome so callers can
    record the attempt uniformly. Never raises — HTTP / signing failures
    surface on the result object.
    """
    start = time.perf_counter()
    try:
        import httpx  # type: ignore
    except ImportError:
        return WebhookDeliveryResult(
            webhook_id=webhook.id,
            status_code=None,
            ok=False,
            error="httpx not installed",
            duration_ms=0,
        )

    # Authoritative SSRF check, at the moment of the request (DNS may have
    # changed since the webhook was created).
    try:
        validate_outbound_url(webhook.url)
    except WebhookUrlError as e:
        return WebhookDeliveryResult(
            webhook_id=webhook.id,
            status_code=None,
            ok=False,
            error=f"refused: {e}",
            duration_ms=0,
        )

    body = json.dumps({"event": event, "data": payload}, default=str).encode()
    timestamp = str(int(time.time()))
    nonce = uuid.uuid4().hex
    signature = _sign(webhook.secret, body, timestamp, nonce)

    headers = {
        "Content-Type": "application/json",
        "X-Webhook-Signature": signature,
        "X-Webhook-Timestamp": timestamp,
        "X-Webhook-Nonce": nonce,
        "X-Webhook-Event": event,
        "X-Webhook-Id": str(webhook.id),
    }
    if webhook.extra_headers:
        headers.update(webhook.extra_headers)

    try:
        # follow_redirects=False is httpx's default; pin it explicitly so a
        # 3xx to an internal host can't bypass the SSRF guard above.
        async with httpx.AsyncClient(timeout=10.0, follow_redirects=False) as client:
            resp = await client.post(webhook.url, content=body, headers=headers)
        return WebhookDeliveryResult(
            webhook_id=webhook.id,
            status_code=resp.status_code,
            ok=resp.is_success,
            error=None if resp.is_success else f"http {resp.status_code}",
            duration_ms=int((time.perf_counter() - start) * 1000),
        )
    except Exception as e:  # noqa: BLE001
        logger.warning("webhook delivery failed: %s -> %s: %s", event, webhook.url, e)
        return WebhookDeliveryResult(
            webhook_id=webhook.id,
            status_code=None,
            ok=False,
            error=str(e),
            duration_ms=int((time.perf_counter() - start) * 1000),
        )


def generate_secret() -> str:
    """Return a 64-hex-char secret suitable for HMAC signing."""
    return uuid.uuid4().hex + uuid.uuid4().hex

File: forge/features/platform/templates/webhooks/node/files/src/services/webhook.service.ts
/**
 * Webhook registry + HMAC-SHA256 delivery for Node/Fastify.
 *
 * In-memory registry in v1 — suitable for dev and single-replica prod. For
 * multi-replica deployments, swap the `registry` Map for a Prisma-backed
 * repository against a `webhooks` table (mirror the Python feature's
 * model shape).
 */

import { createHmac, randomBytes, randomUUID } from "node:crypto";

export interface Webhook {
	id: string;
	name: string;
	url: string;
	secret: string;
	events: string[];
	is_active: boolean;
	extra_headers: Record<string, string> | null;
	created_at: string;
}

export interface WebhookCreate {
	name: string;
	url: string;
	events?: string[];
	extra_headers?: Record<string, string>;
}

export interface DeliveryResult {
	webhook_id: string;
	status_code: number | null;
	ok: boolean;
	error: string | null;
	duration_ms: number;
}

const registry = new Map<string, Webhook>();

export class WebhookUrlError extends Error {}

/**
 * Reject webhook targets that point at internal/non-public hosts (SSRF) or use
 * a non-http(s) scheme. Mirrors the Python feature's `validate_outbound_url`.
 *
 * Host literals in loopback / link-local (incl. the 169.254.169.254 cloud
 * metadata endpoint) / RFC1918 private ranges are blocked. DNS names are not
 * resolved here (no sync resolver in the fetch path); pair this pre-flight
 * check with `redirect: "manual"` below so a 3xx to an internal host cannot
 * slip past it.
 */
function validateOutboundUrl(rawUrl: string): void {
	let parsed: URL;
	try {
		parsed = new URL(rawUrl);
	} catch {
		throw new WebhookUrlError(`invalid webhook URL: ${rawUrl}`);
	}
	const scheme = parsed.protocol.toLowerCase();
	if (scheme !== "https:" && scheme !== "http:") {
		throw new WebhookUrlError(`unsupported URL scheme ${scheme}; use https`);
	}
	// `hostname` strips brackets from IPv6 literals; lowercase for name checks.
	const host = parsed.hostname.toLowerCase();
	if (!host) {
		throw new WebhookUrlError("webhook URL has no host");
	}
	if (isBlockedHost(host)) {
		throw new WebhookUrlError(`${host} is a non-public address; refused`);
	}
}

function isBlockedHost(host: string): boolean {
	if (host === "localhost" || host.endsWith(".localhost")) return true;
	// IPv6 literals: loopback (::1), unspecified (::), link-local (fe80::/10),
	// unique-local (fc00::/7), and IPv4-mapped loopback (::ffff:127.0.0.1).
	if (host.includes(":")) {
		if (host === "::1" || host === "::") return true;
		if (host.startsWith("fe8") || host.startsWith("fe9")) return true;
		if (host.startsWith("fea") || host.startsWith("feb")) return true;
		if (host.startsWith("fc") || host.startsWith("fd")) return true;
		const mapped = host.split(":").pop() ?? "";
		if (mapped.includes(".")) return isBlockedIPv4(mapped);
		return false;
	}
	if (isBlockedIPv4(host)) return true;
	return false;
}

function isBlockedIPv4(host: string): boolean {
	const parts = host.split(".");
	if (parts.length !== 4) return false;
	const octets = parts.map((p) => Number(p));
	if (octets.some((n) => !Number.isInteger(n) || n < 0 || n > 255)) {
		return false;
	}
	const [a, b] = octets;
	if (a === 127) return true; // loopback 127.0.0.0/8
	if (a === 10) return true; // RFC1918 10.0.0.0/8
	if (a === 192 && b === 168) return true; // RFC1918 192.168.0.0/16
	if (a === 172 && b >= 16 && b <= 31) return true; // RFC1918 172.16.0.0/12
	if (a === 169 && b === 254) return true; // link-local 169.254.0.0/16 (metadata)
	if (a === 0) return true; // 0.0.0.0/8 unspecified
	return false;
}

function generateSecret(): string {
	return randomBytes(32).toString("hex");
}

function matchesEvent(webhook: Webhook, event: string): boolean {
	if (!webhook.events || webhook.events.length === 0) return true;
	// glob-style suffix matching: "item.*" matches "item.created"
	return webhook.events.some((pattern) => {
		if (pattern === event) return true;
		if (pattern.endsWith("*")) {
			return event.startsWith(pattern.slice(0, -1));
		}
		return false;
	});
}

function sign(
	secret: string,
	timestamp: string,
	nonce: string,
	body: Buffer,
): string {
	return createHmac("sha256", secret)
		.update(timestamp)
		.update(".")
		.update(nonce)
		.update(".")
		.update(body)
		.digest("hex");
}

export function listWebhooks(): Webhook[] {
	return Array.from(registry.values()).sort((a, b) =>
		b.created_at.localeCompare(a.created_at),
	);
}

export function createWebhook(data: WebhookCreate): Webhook {
	const id = crypto.randomUUID();
	const webhook: Webhook = {
		id,
		name: data.name,
		url: data.url,
		secret: generateSecret(),
		events: data.events ?? [],
		is_active: true,
		extra_headers: data.extra_headers ?? null,
		created_at: new Date().toISOString(),
	};
	registry.set(id, webhook);
	return webhook;
}

export function getWebhook(id: string): Webhook | null {
	return registry.get(id) ?? null;
}

export function deleteWebhook(id: string): boolean {
	return registry.delete(id);
}

export async function deliver(
	webhook: Webhook,
	event: string,
	payload: unknown,
): Promise<DeliveryResult> {
	const start = performance.now();

	// Fast pre-flight SSRF reject (scheme + host literal). Paired with
	// `redirect: "manual"` below so a 3xx to an internal host can't bypass it.
	try {
		validateOutboundUrl(webhook.url);
	} catch (err: any) {
		return {
			webhook_id: webhook.id,
			status_code: null,
			ok: false,
			error: `refused: ${String(err?.message ?? err)}`,
			duration_ms: Math.round(performance.now() - start),
		};
	}

	const timestamp = Math.floor(Date.now() / 1000).toString();
	const nonce = randomUUID().replace(/-/g, "");
	const body = Buffer.from(
		JSON.stringify({ event, data: payload }),
		"utf-8",
	);
	const signature = sign(webhook.secret, timestamp, nonce, body);

	const headers: Record<string, string> = {
		"Content-Type": "application/json",
		"X-Webhook-Signature": signature,
		"X-Webhook-Timestamp": timestamp,
		"X-Webhook-Nonce": nonce,
		"X-Webhook-Event": event,
		"X-Webhook-Id": webhook.id,
		...(webhook.extra_headers ?? {}),
	};

	const controller = new AbortController();
	const timeout = setTimeout(() => controller.abort(), 10000);
	try {
		const resp = await fetch(webhook.url, {
			method: "POST",
			headers,
			body,
			// Do not auto-follow 3xx: a redirect to an internal host would
			// bypass the validateOutboundUrl pre-flight check above.
			redirect: "manual",
			signal: controller.signal,
		});
		return {
			webhook_id: webhook.id,
			status_code: resp.status,
			ok: resp.ok,
			error: resp.ok ? null : `http ${resp.status}`,
			duration_ms: Math.round(performance.now() - start),
		};
	} catch (err: any) {
		return {
			webhook_id: webhook.id,
			status_code: null,
			ok: false,
			error: String(err?.message ?? err),
			duration_ms: Math.round(performance.now() - start),
		};
	} finally {
		clearTimeout(timeout);
	}
}

export async function fireEvent(
	event: string,
	payload: unknown,
): Promise<DeliveryResult[]> {
	const results: DeliveryResult[] = [];
	for (const webhook of registry.values()) {
		if (webhook.is_active && matchesEvent(webhook, event)) {
			results.push(await deliver(webhook, event, payload));
		}
	}
	return results;
}
